File: typescript/src/giftcard/domain/policies/confirmTopUpPolicy.ts
// Policy che reagisce agli esiti del pagamento e richiede la conferma della ricarica.
// Violazione cross-BC didattica: GiftCard non dovrebbe importare tipi da Payment.

import type { Policy } from '@/common/domain/model/policy.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import type { PaymentResultEvent } from '@/payment/domain/events/paymentResultEvents.js';
import { match } from 'ts-pattern';
import { ConfirmTopUp, confirmTopUp } from '../../application/commands/confirmTopUp.js';
import { GiftCardId } from '../giftcard/giftCardId.js';

export class ConfirmTopUpPolicy implements Policy<PaymentResultEvent, ConfirmTopUp> {
  evaluate(event: PaymentResultEvent): ConfirmTopUp | null {
    return match(event)
      .with(
        { kind: 'PaymentAccepted' },
        (accepted) => confirmTopUp(new GiftCardId(Uuid.fromString(accepted.clientReference.value)), accepted.amount)
      )
      .with({ kind: 'PaymentRejected' }, () => null)
      .with({ kind: 'PaymentExpired' }, () => null)
      .exhaustive();
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/domain/policies/ConfirmTopUpPolicy.java
package it.giannibombelli.wsc2026.giftcard.domain.policies;

import it.giannibombelli.wsc2026.common.domain.model.Policy;
import it.giannibombelli.wsc2026.common.utils.Require;
import it.giannibombelli.wsc2026.giftcard.application.commands.ConfirmTopUp;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;

import java.util.UUID;

import static it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents.*;


public final class ConfirmTopUpPolicy implements Policy<PaymentResultEvents, ConfirmTopUp> {
    @Override
    public ConfirmTopUp evaluate(PaymentResultEvents event) {
        Require.requireArgument(event, "Payment event");
        return switch (event) {
            case PaymentAccepted it ->
                new ConfirmTopUp(new GiftCardId(it.clientReference().value()), it.amount());
            case PaymentRejected ignored -> null;
            case PaymentExpired ignored -> null;
        };
    }
}
